pylint/checkers/unicode.py

```diff
@@ -0,0 +1,701 @@
+
+class NoneAsciiNamesChecker(BaseChecker):
+    """A very strict version of the name checker that only allows ASCII for everything!
+
+    If your programming guideline defines that your are programming in English,
+    then there should be no need for non ASCII characters.
+    Everybody else can simply not use this test.
+    """
+
+    __implements__ = interfaces.IAstroidChecker
+    priority = -1
+
+    msgs = {
+        "C0144": (
+            '%s name "%s" contains a non-ASCII unicode character',
+            "non-ascii-name",
+            "Used when the name contains at least one non-ASCII unicode character.",
+        ),
+    }
+
+    name = "NonASCII-Checker"
+
+    def __init__(self, linter):
+        super().__init__(linter)
+        self._non_ascii_rgx_compiled = re.compile("[^A-Za-z0-9_]")
+
+    def _raise_name_warning(
+        self,
+        node: nodes.NodeNG,
+        node_type: str,
+        name: str,
+        confidence: interfaces.Confidence,
+    ) -> None:
+        type_label = HUMAN_READABLE_TYPES[node_type]
+        args = (type_label.capitalize(), name)
+
+        self.add_message("non-ascii-name", node=node, args=args, confidence=confidence)
+        self.linter.stats.increase_bad_name(node_type, 1)
+
+    def _check_name(
+        self,
+        node_type: str,
+        names: Union[str, List[str]],
+        node: Union[nodes.NodeNG, _AsciiOnlyCheckedNode],
+        confidence: interfaces.Confidence = interfaces.HIGH,
+    ):
+        """check for a name if it's using non ASCII and set message if it does
+
+        Also set the dynamic attribute ``_is_ascii_only`` that is used to
+        determine if a node has been already checked, so we don't have to handle
+        too many edge cases.
+        """
+        if not isinstance(names, list):
+            names = [names]
+
+        current_state = getattr(node, "_is_ascii_only", True)
+        for name in names:
+            if name is None:
+                # For some nodes i.e. *kwargs from a dict, the name will be empty
+                continue
+            non_ascii_match = self._non_ascii_rgx_compiled.match(name)
+            if not (Py37Str(name).isascii() and non_ascii_match is None):
+                # Note that we require the `.isascii` method as it is fast and
+                # handles the complexities of unicode, so we can use simple regex.
+                self._raise_name_warning(
+                    node,
+                    node_type,
+                    name,
+                    confidence,
+                )
+                current_state = False
+
+        node._is_ascii_only = current_state  # pylint: disable=protected-access
+
+    def _recursive_check_names(self, args):
+        """check names in a possibly recursive list <arg>"""
+        for arg in args:
+            if isinstance(arg, nodes.AssignName):
+                self._check_name("argument", arg.name, arg)
+            else:
+                self._recursive_check_names(arg.elts)
```

Comment:
This is uncovered, can we make a test case for it?

Reply:
This is based on the original name checker, which I simplified. I don't know at test case for that as I am lacking the deeper AST knowledge

Reply:
👍